### PR TITLE
Deprecate colon prefixed route template placeholders.

### PR DESCRIPTION
**Pull request by ADmad**
Braced placeholders should be used instead.

### CODE REVIEW

File: src/Routing/Route/Route.php [hunk 1]
```diff
@@ -852,12 +858,26 @@ public function staticPath(): string
     {
         $routeKey = strpos($this->template, '{');
         if ($routeKey !== false && strpos($this->template, '}') !== false) {
-            return substr($this->template, 0, $routeKey);
+            $parts = explode('/', ltrim($this->template, '/'));
```
Comment (ADmad) [inline src/Routing/Route/Route.php:861]: This additional work was required to support templates like `'/api/{/{action}/*'`.

Someone actually using such a template seems quite unlikely but our test suite has a test for `'/api/{/:action/*'`, it needed to be supported with the braced placeholder too.

File: src/Routing/Route/Route.php [hunk 2]
```diff
@@ -319,8 +319,14 @@ protected function _writeRoute(): void
         if (strpos($route, '{') !== false && strpos($route, '}') !== false) {
             preg_match_all('/\{([a-z][a-z0-9-_]*)\}/i', $route, $namedElements);
         } else {
-            preg_match_all('/:([a-z0-9-_]+(?<![-_]))/i', $route, $namedElements);
+            $hasMatches = preg_match_all('/:([a-z0-9-_]+(?<![-_]))/i', $route, $namedElements);
             $this->braceKeys = false;
+            if ($hasMatches) {
+                deprecationWarning(
+                    'Colon prefixed route placeholders are deprecated.'
```
Comment (othercorey) [inline src/Routing/Route/Route.php:326]: Let's add a matching example of colon prefixed for foo.

File: tests/TestCase/Routing/Route/RouteTest.php [hunk 3]
```diff
@@ -39,16 +39,24 @@ public function setUp(): void
         Configure::write('Routing', ['prefixes' => []]);
     }
 
+    public function testDeprecationWarning()
```
Comment (othercorey) [inline tests/TestCase/Routing/Route/RouteTest.php:42]:
```suggestion
    public function testDeprecatedPlaceholders()
```

File: tests/TestCase/Routing/Route/RouteTest.php [hunk 4]
```diff
@@ -39,16 +39,24 @@ public function setUp(): void
...
+    {
+        $this->deprecated(function () {
+            $route = new Route('/:controller');
```
Comment (othercorey) [inline tests/TestCase/Routing/Route/RouteTest.php:45]:
```suggestion
            $route = new Route('/:controller/:action/:id');
```

File: tests/TestCase/Routing/Route/RouteTest.php (hunk 3 above)
Comment (ADmad) [reply tests/TestCase/Routing/Route/RouteTest.php:42]: Don't be so pedantic :stuck_out_tongue:.

File: src/Routing/Route/Route.php (hunk 1 above)
Comment (markstory) [reply src/Routing/Route/Route.php:861]: This looks like a sizable amount of code for something that could lead to ambiguous results. We could require that any `{` in URL templates require an escape character like `\{` to help clarify where the route parameters are.

Comment (ADmad) [reply src/Routing/Route/Route.php:861]: Yeah that sounds reasonable.

Comment (othercorey) [reply src/Routing/Route/Route.php:861]: I remember voicing a concern about escaping unrelated `{` characters in the text/log placeholders a long time ago. Did we do the same thing there?

File: src/Routing/Route/Route.php [hunk 5]
```diff
@@ -850,14 +861,22 @@ protected function _writeUrl(array $params, array $pass = [], array $query = [])
      */
     public function staticPath(): string
     {
-        $routeKey = strpos($this->template, '{');
-        if ($routeKey !== false && strpos($this->template, '}') !== false) {
-            return substr($this->template, 0, $routeKey);
+        $matched = preg_match(
```
Comment (ADmad) [inline src/Routing/Route/Route.php:864]: @markstory I replaced the array juggling with a single preg_match. This shouldn't add too much overhead.

Comment (markstory) [reply src/Routing/Route/Route.php:864]: Much simpler :clap:

**Pull request merged**